src/main/java/io/temporal/failure/ApplicationFailure.java

```diff
@@ -51,49 +51,47 @@
  */
 public final class ApplicationFailure extends TemporalFailure {
   private final String type;
-  private final Value details;
+  private final Values details;
   private boolean nonRetryable;
 
   /**
+   * New ApplicationFailure with {@link #isNonRetryable()} flag set to false. Note that this
+   * exception still can be retried by the service if its type is included into doNotRetry property
+   * of the correspondent retry policy.
+   *
    * @param message optional error message
    * @param type optional error type that is used by {@link
    *     io.temporal.common.RetryOptions#addDoNotRetry(String...)}.
    * @param details optional details about the failure. They are serialized using the same approach
-   *     as arguments and results and can be accessed through {@link #getDetails()}
-   * @param cause failure cause. Each element of the cause chain is converted to ApplicationFailure
-   *     if it doesn't extend {@link TemporalFailure}.
+   *     as arguments and results.
    */
-  public ApplicationFailure(String message, String type, Object details, Exception cause) {
-    this(message, type, new EncodedValue(details), false, cause);
+  public static ApplicationFailure newFailure(String message, String type, Object... details) {
+    return new ApplicationFailure(message, type, false, new EncodedValues(details), null);
   }
 
   /**
+   * New ApplicationFailure with {@link #isNonRetryable()} flag set to true.
+   *
+   * <p>It means that this exception is not going to be retried even if it is not included into
+   * retry policy doNotRetry list.
+   *
    * @param message optional error message
-   * @param type optional error type that is used by {@link
-   *     io.temporal.common.RetryOptions#addDoNotRetry(String...)}.
+   * @param type optional error type
    * @param details optional details about the failure. They are serialized using the same approach
    *     as arguments and results.
    */
-  public ApplicationFailure(String message, String type, Object details) {
-    this(message, type, new EncodedValue(details), false, null);
-  }
-
-  /**
-   * @param message optional error message
-   * @param type optional error type that is used by {@link
-   *     io.temporal.common.RetryOptions#addDoNotRetry(String...)}.
-   */
-  public ApplicationFailure(String message, String type) {
-    this(message, type, new EncodedValue(null), false, null);
+  public static ApplicationFailure newNonRetryableFailure(
+      String message, String type, Object... details) {
+    return new ApplicationFailure(message, type, true, new EncodedValues(details), null);
   }
 
-  /** * @param message optional error message */
-  public ApplicationFailure(String message) {
-    this(message, null);
+  static ApplicationFailure newFromValues(
+      String message, String type, boolean nonRetryable, Values details, Throwable cause) {
```

Comment:
[Nit], having boolean names that are negative (nonRetryable vs retryable) can be kind of confusing

Reply:
I almost always follow the pattern that the default value should be the default value for the type. So the default of the boolean type is false. So the field is called nonRetryable as its default is false. But I agree that it looks weird in the parameter name.